WT-6330 Perform RTS on the history store as a final step

kommiharibabu opened the pull request:
As the history store is checkpointed at the last step of
all checkpoint, during this time it is possible that newer
updates are written to the history from the data store tables
by eviction process that are not present in the data store.
Due to this it is possible that sometimes history store may
have newer updates that are not present in the data store
checkpoint. If any backup/unclean shutdown uses this inconsistent
checkpoint it is possible that history store may have data newer
than the stable timestamp and can cause data visibility issues.

To avoid this problem, while recovering the database, perform
rollback to stable on the history store at the end.

bvpvamsikrishna commented:
@kommiharibabu, could you post the patch build link here.

kommiharibabu commented:
Evergreen patch build link - https://evergreen.mongodb.com/version/5ed6fcc03066150e6c0e47c6
I see two failures that happen on the patch. Need to check and confirm

kommiharibabu commented:
I checked all the failures and there is no issue from the logs and they seem stopped in the middle. I restarted them and all are passed.

bvpvamsikrishna commented on src/txn/txn_rollback_to_stable.c:
```diff
@@ -998,6 +1016,74 @@ __rollback_to_stable_btree_hs_truncate(WT_SESSION_IMPL *session, uint32_t btree_
+     * 2. The checkpoint durable start/stop timestamp is greater than the rollback timestamp.
+     * 3. There is no durable timestamp in any checkpoint.
+     */
+    if (S2BT(session)->modified || max_durable_ts > rollback_timestamp || !durable_ts_found) {
```
for which case do we require, the condition `!durable_ts_found` only?  I think that condition is redundant.

kommiharibabu replied:
Yes, those all conditions are not required. Just the timestamp check is enough. Fixed.

kommiharibabu commented on src/txn/txn_rollback_to_stable.c:
```diff
@@ -998,6 +1016,74 @@ __rollback_to_stable_btree_hs_truncate(WT_SESSION_IMPL *session, uint32_t btree_
 
+/*
+ * __rollback_to_stable_hs_final_pass --
+ *     Perform rollback to stable on the history store to remove any updates more than stable
```
Fixed.

kommiharibabu commented on src/txn/txn_rollback_to_stable.c:
```diff
@@ -355,14 +355,28 @@ __rollback_abort_row_ondisk_kv(
+    if (WT_IS_HS(S2BT(session))) {
+        if (vpack->tw.durable_stop_ts > rollback_timestamp) {
+            __wt_verbose(session, WT_VERB_RTS,
+              "hs update aborted with start durable timestamp: %s, commit timestamp: %s, "
```
Instead of adding start/stop to every timestamp, I changed something like the following, how do you think?
`hs update aborted with start durable/commit timestamp: %s, %s`

kommiharibabu commented on dist/stat_data.py:
```diff
@@ -651,6 +651,7 @@ def __init__(self, name, desc, flags=''):
     TxnStat('txn_rollback', 'transactions rolled back'),
     TxnStat('txn_rts', 'rollback to stable calls'),
     TxnStat('txn_rts_hs_removed', 'rollback to stable updates removed from history store'),
+    TxnStat('txn_rts_hs_keys_larger_than_datastore_removed', 'rollback to stable updates more than data store removed from history store'),
```
Changed to `txn_rts_sweep_hs_keys`.

kommiharibabu commented:
@bvpvamsikrishna I fixed the random restart failure scenario. This is happening because of evicting all the locally read pages with fix given in WT-6350 led to split the page while the rollback to stable is in progress. I fixed it by not splitting the locally read pages while evicting them.

@luke-pearson Can you review the code changes. 
I will add a test case based on the timing stress that gets introduced in WT-6371
